File: packages/guiobj/c/src/display.c
/*
 * jMax
 * Copyright (C) 1994, 1995, 1998, 1999 by IRCAM-Centre Georges Pompidou, Paris, France.
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * See file LICENSE for further informations on licensing terms.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * 
 * Based on Max/ISPW by Miller Puckette.
 *
 * Authors: Maurizio De Cecco, Francois Dechelle, Enzo Maggi, Norbert Schnell.
 *
 */

#include <string.h>
#include "fts.h"

#define STRING_SIZE 256

typedef struct {
  fts_object_t o;
  fts_atom_t a;
  char string[STRING_SIZE];
  fts_alarm_t alarm;
  double period;
  int gate;
  int pending;
} display_t;

static void
append_blank_and_atom(char *str, const fts_atom_t *a)
{
  int n = strlen(str);
  char *s = str + n;

  if(fts_is_int(a))
    snprintf(s, STRING_SIZE - n, " %d", fts_get_int(a));
  else if(fts_is_float(a))
    snprintf(s, STRING_SIZE - n, " %g", fts_get_float(a));
  else if(fts_is_symbol(a))
    snprintf(s, STRING_SIZE - n, " %s", fts_symbol_name(fts_get_symbol(a)));
  else
    snprintf(s, STRING_SIZE - n, " %s", "?");
}

static void
append_atom(char *str, const fts_atom_t *a)
{
  int n = strlen(str);
  char *s = str + n;

  if(fts_is_int(a))
    snprintf(s, STRING_SIZE - n, "%d", fts_get_int(a));
  else if(fts_is_float(a))
    snprintf(s, STRING_SIZE - n, "%g", fts_get_float(a));
  else if(fts_is_symbol(a))
    snprintf(s, STRING_SIZE - n, "%s", fts_symbol_name(fts_get_symbol(a)));
  else
    snprintf(s, STRING_SIZE - n, "%s", "?");
}

/************************************************************
 *
 *  send to client with time gate
 *
 */
static void
display_deliver(display_t *this)
{
  if(this->gate)
    {
      this->pending = 0;
      this->gate = 0;

      fts_client_send_message((fts_object_t *)this, fts_s_set, 1, &this->a);

      fts_alarm_set_delay(&this->alarm, this->period);
      fts_alarm_arm(&this->alarm);
    }
  else
    this->pending = 1;
}

static void
display_alarm(fts_alarm_t *alarm, void *o)
{
  display_t * this = (display_t *)o;

  if(this->pending)
    {
      this->gate = 0;
      this->pending = 0;

      fts_client_send_message(o, fts_s_set, 1, &this->a);

      fts_alarm_set_delay(&this->alarm, this->period);
      fts_alarm_arm(&this->alarm);
    }
  else
    this->gate = 1;
}

/************************************************************
 *
 *  object
 *
 */
static void
display_init(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  display_t * this = (display_t *)o;

  fts_set_string(&this->a, this->string);
  fts_alarm_init(&this->alarm, 0, display_alarm, (void *)this);

  this->period = 50.0;
  this->gate = 1;
  this->pending = 0;
}

static void
display_int(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  display_t * this = (display_t *)o;

  sprintf(this->string, "%d", fts_get_int(at));

  display_deliver(this);
}

static void
display_float(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  display_t * this = (display_t *)o;

  sprintf(this->string, "%g", fts_get_float(at));

  display_deliver(this);
}

static void
display_symbol(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  display_t * this = (display_t *)o;

  sprintf(this->string, "%s", fts_symbol_name(fts_get_symbol(at)));

  display_deliver(this);
}

static void 
display_list(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  display_t * this = (display_t *)o;
  int i;

  this->string[0] = '\0';
  append_atom(this->string, at);

  for(i=1; i<ac; i++)
    append_blank_and_atom(this->string, at + i);

  display_deliver(this);
}

static void 
display_anything(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  display_t * this = (display_t *)o;
  int i;

  sprintf(this->string, "<%s>", fts_symbol_name(s));

  for(i=0; i<ac; i++)
    append_blank_and_atom(this->string, at + i);

  display_deliver(this);
}

static fts_status_t 
display_instantiate(fts_class_t *cl, int ac, const fts_atom_t *at)
{
  fts_class_init(cl, sizeof(display_t), 1, 0, 0);

  fts_method_define_varargs(cl, fts_SystemInlet, fts_s_init, display_init);

  fts_method_define_varargs(cl, 0, fts_s_int, display_int);
  fts_method_define_varargs(cl, 0, fts_s_float, display_float);
  fts_method_define_varargs(cl, 0, fts_s_symbol, display_symbol);
  fts_method_define_varargs(cl, 0, fts_s_list, display_list);
  fts_method_define_varargs(cl, 0, fts_s_anything, display_anything);

  return fts_Success;
}

void 
display_config(void)
{
  fts_class_install(fts_new_symbol("display"), display_instantiate);
}

File: java/src/ircam/jmax/editors/patcher/objects/Display.java
//
// jMax
// Copyright (C) 1994, 1995, 1998, 1999 by IRCAM-Centre Georges Pompidou, Paris, France.
// 
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
// 
// See file LICENSE for further informations on licensing terms.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
// 
// Based on Max/ISPW by Miller Puckette.
//
// Authors: Maurizio De Cecco, Francois Dechelle, Enzo Maggi, Norbert Schnell.
// 

package ircam.jmax.editors.patcher.objects;

import java.awt.*;
import java.awt.event.*;
import java.util.*;

import ircam.jmax.fts.*;
import ircam.jmax.editors.patcher.*;
import ircam.jmax.editors.patcher.menus.*;
import ircam.jmax.editors.patcher.interactions.*;

//
// The display graphic object.
//

class Display extends GraphicObject implements FtsMessageListener
{
  String display = null; // the display content
  int minWidth = 0; //

  public Display(ErmesSketchPad theSketchPad, FtsObject theFtsObject)
  {
    super(theSketchPad, theFtsObject);

    redefine("     ");

    setFont(getFont());
  }

  // ----------------------------------------
  // ``Args'' property
  // ----------------------------------------

  public String getArgs()
  {
    return display;
  }

  public void redefine( String text) 
  {
    int w = getFontMetrics().stringWidth(text) + 4;
    int h = getFontMetrics().getHeight() + 4;

    display = text;

    if(w < minWidth)
      super.setWidth(minWidth);
    else
      super.setWidth( w);

    super.setHeight( h);
  }

  // Set the text when FTS change the display content
  public void messageChanged(String text)
  {
    redraw();
    redefine( text);
    redraw();
  }

  public void setWidth( int theWidth) 
  {
  }

  public void setHeight( int theHeight)
  {
  }

  public void gotSqueack(int squeack, Point mouse, Point oldMouse)
  {          
  }

  // redefined from base class
  public void setFont( Font theFont)
  {
    super.setFont( theFont);

    int w = getFontMetrics().stringWidth(display) + 4;
    int h = getFontMetrics().getHeight() + 4;

    minWidth = h;

    if(w < minWidth)
      super.setWidth(minWidth);
    else
      super.setWidth( w);

    super.setHeight( h);

    redraw();
  }

  // ----------------------------------------
  // ``TextBackground'' property
  // ----------------------------------------

  public Color getTextForeground()
  {
    return Color.black;
  }

  public Color getTextBackground()
  {
    if (itsSketchPad.isLocked())
      return Color.white;
    else
      {
	if (isSelected())
	  return Color.gray;
	else
	  return itsSketchPad.getBackground();
      }
  }

  public void paint(Graphics g) 
  {
    int x = getX();
    int y = getY();
    int w = getWidth();
    int h = getHeight();

    if(itsSketchPad.isLocked())
      {
	g.setColor( Color.white);
	g.fillRect( x + 1, y + 1, w - 2, h - 2);

	g.setColor( Color.black);
	//g.drawLine( x, y, x + w - 1, y);
	g.drawLine( x, y, x + minWidth, y);
	g.drawLine( x, y, x, y + h/4);
	
	paintInlets(g);
	
	g.setColor( Color.gray);
	g.setFont(getFont());
	g.drawString(display, x + 2, y + h - getFontMetrics().getDescent() - 2);
      }
    else
      {
	if (isSelected())
	  g.setColor(Color.gray);
	else
	  g.setColor(itsSketchPad.getBackground());

	g.fill3DRect( x, y + 1, w, h - 1, true); 

	g.setColor( Color.black);
	//g.drawLine( x, y, x + w - 1, y);
	g.drawLine( x, y, x + minWidth, y);
	g.drawLine( x, y, x, y + h/4);

	paintInlets(g);
	
	g.setFont(getFont());
	g.drawString(display, x + 2, y + h - getFontMetrics().getDescent() - 2);
      }
  }

  public void updatePaint(Graphics g) 
  {
    int x = getX();
    int y = getY();
    int w = getWidth();
    int h = getHeight();

    g.setFont(getFont());
    g.drawString(display, x + 2, y + h - getFontMetrics().getDescent() - 2);
  }

  public void popUpUpdate(boolean onInlet, boolean onOutlet, SensibilityArea area)
  {
    super.popUpUpdate(onInlet, onOutlet, area);
    TextPopUpMenu.update(this);
    ObjectPopUp.addMenu(TextPopUpMenu.getInstance());
  }
  public void popUpReset()
  {
    super.popUpReset();
    ObjectPopUp.removeMenu(TextPopUpMenu.getInstance());
  }
}
